src/pretty.rs

```diff
@@ -1,64 +1,56 @@
-                    content,
-                    single_line_no_indent,
-                    ..
-                } => {
-                    if *single_line_no_indent {
-                        RcDoc::hardline().append(RcDoc::text(content.trim_end()))
-                    } else {
-                        RcDoc::text(content.trim_end())
-                    }
+    for atom in atoms {
+        let extra = match atom {
+            Atom::Blankline => format!("\n\n{}", " ".repeat(indent_level)),
+            Atom::Empty => String::new(),
+            &Atom::Hardline => format!("\n{}", " ".repeat(indent_level)),
+            Atom::IndentEnd => {
+                if indent_offset > indent_level {
+                    return Err(FormatterError::Query(
+                        "Trying to close an unopened indentation block".into(),
+                        None,
+                    ));
+                } else {
+                    indent_level -= indent_offset;
+                    String::new()
                 }
-                Atom::Literal(s) => RcDoc::text(s),
-                Atom::MultilineOnlyLiteral { .. } => unreachable!(),
-                Atom::IndentEnd => unreachable!(),
-                Atom::IndentStart => {
-                    *i += 1;
-                    *indent_level += indent_offset;
-                    let res = atoms_to_doc(i, atoms, indent_offset, indent_level);
-                    *indent_level -= indent_offset;
-                    res
+            }
+            Atom::IndentStart => {
+                indent_level += indent_offset;
+                String::new()
+            }
+            Atom::Leaf {
+                content,
+                single_line_no_indent,
+                ..
+            } => {
+                if *single_line_no_indent {
+                    // The line break after the content has been previously added
+                    // as a `Hardline` in the atom stream.
+                    format!("\n{}", content.trim_end())
+                } else {
+                    content.trim_end().to_string()
                 }
-                Atom::Softline { .. } => unreachable!(),
-                Atom::Space => RcDoc::space(),
-                Atom::DeleteBegin => unreachable!(),
-                Atom::DeleteEnd => unreachable!(),
-                Atom::ScopedSoftline { .. } => unreachable!(),
-            });
-        }
-        *i += 1;
+            }
+            Atom::Literal(s) => s.to_string(),
+            Atom::Space => " ".to_string(),
+            // All other atom kinds should have been post-processed at that point
+            other => {
+                return Err(FormatterError::Internal(
+                    format!(
+                        "Found atom that should have been removed before rendering: {:?}",
+                        other
+                    ),
+                    None,
+                ))
+            }
+        };
+        buffer.push_str(extra.as_str());
```

Comment:
I don't think it's necessary to make this change, but... `String` implements `std::fmt::Write`, so rather than constructing a new string and appending it to the buffer, you could `write!` directly into the buffer.

```rust
use std::fmt::write;

fn main() {
    let mut a: String = String::new();
    write!(a, "Hello, world!").unwrap();
    println!("{a}");
}
```

Reply:
Well spotted, I've made the modification in the new version.